project-minimization-plugin/src/main/kotlin/org/plan/research/minimization/plugin/services/BenchmarkingService.kt

```diff
@@ -0,0 +1,151 @@
+package org.plan.research.minimization.plugin.services
+
+import org.plan.research.minimization.plugin.benchmark.BenchmarkConfig
+import org.plan.research.minimization.plugin.benchmark.BenchmarkProject
+import org.plan.research.minimization.plugin.benchmark.BenchmarkResultAdapter
+import org.plan.research.minimization.plugin.benchmark.BuildSystemType
+import org.plan.research.minimization.plugin.errors.MinimizationError
+import org.plan.research.minimization.plugin.model.FileLevelStage
+import org.plan.research.minimization.plugin.model.state.DDStrategy
+import org.plan.research.minimization.plugin.model.state.HierarchyCollectionStrategy
+import org.plan.research.minimization.plugin.settings.MinimizationPluginSettings
+
+import arrow.core.Either
+import arrow.core.None
+import arrow.core.Option
+import arrow.core.raise.option
+import com.charleskorn.kaml.Yaml
+import com.intellij.ide.impl.ProjectUtil
+import com.intellij.openapi.application.EDT
+import com.intellij.openapi.components.Service
+import com.intellij.openapi.components.service
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.project.ProjectManager
+import com.intellij.openapi.project.guessProjectDir
+import com.intellij.openapi.vfs.findFile
+import com.intellij.openapi.vfs.readText
+import com.intellij.openapi.vfs.toNioPathOrNull
+
+import java.nio.file.Path
+
+import kotlin.io.path.Path
+import kotlin.io.path.readText
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.flow.asFlow
+import kotlinx.coroutines.flow.filter
+import kotlinx.coroutines.flow.filterNotNull
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+
+@Service(Service.Level.PROJECT)
+class BenchmarkingService(private val rootProject: Project, private val cs: CoroutineScope) {
+    fun benchmark(adapter: BenchmarkResultAdapter) = cs.launch {
+        val config = readConfig().getOrNull()
```

Comment:
We need to support cancellation + show progress of this action, see how it works in minimize project action